### ISSUE TITLE
Inexact reveal for a generic function with a bound and covariant type variable

### ISSUE THREAD

**Issue by tyralla** [python/mypy#10628]
I've discussed the following problem with @Pyprohly and @A5rocks on Gitter.  Still, not sure if it's a bug or happens for good reasons.

The following is a working example dealing with random numbers in a parallel Scalar-Vector class hierarchy.

Mypy raises an error in the last line due to revealing `object`.   Shouldn't it reveal `Vector[Scalar]`?  At least, it doesn't complain about using `VectorNormal` and `VectorGamma` instances as `Vector[Scalar]` subtypes.

```python
from random import gammavariate, normalvariate, uniform
from typing import Generic, Tuple, TypeVar

T = TypeVar("T")

def select_randomly(x: T, y: T) -> T:
    return x if uniform(0.0, 1.0) > 0.5 else y

class Scalar:
    _value: float
    @property
    def value(self) -> float:
        return self._value

class ScalarNormal(Scalar):
    def __init__(self) -> None:
        self._value = normalvariate(1.0, 2.0)

class ScalarGamma(Scalar):
    def __init__(self) -> None:
        self._value = gammavariate(1.0, 2.0)

ScalarType = TypeVar("ScalarType", bound=Scalar, covariant=True)

class Vector(Generic[ScalarType]):
    _scalars: Tuple[ScalarType, ScalarType]
    @property
    def maximum(self) -> ScalarType:
        if self._scalars[0].value > self._scalars[1].value:
            return self._scalars[0]
        return self._scalars[1]

class VectorNormal(Vector[ScalarNormal]):
    def __init__(self) -> None:
        self._scalars = ScalarNormal(), ScalarNormal()

class VectorGamma(Vector[ScalarGamma]):
    def __init__(self) -> None:
        self._scalars = ScalarGamma(), ScalarGamma()

vn = VectorNormal()
vg = VectorGamma()
v1: Vector[Scalar] = vn
v2: Vector[Scalar] = vg
select_randomly(vn, vg).maximum.value  # error: "object" has no attribute "maximum"
```

**Comment by hauntsaninja** [comment]
It's a bug, fixed by #9994
